core-api/src/test/java/com/optimizely/ab/event/internal/EventBuilderTest.java

```diff
@@ -85,25 +82,93 @@ public EventBuilderTest(int datafileVersion,
+                "value");
+        com.optimizely.ab.event.internal.payload.Attribute userAgentFeature = new com.optimizely.ab.event.internal.payload.Attribute(
+                ReservedAttributeKey.USER_AGENT_ATTRIBUTE.toString(),
+                ReservedAttributeKey.USER_AGENT_ATTRIBUTE.toString(),
+                com.optimizely.ab.event.internal.payload.Attribute.CUSTOM_ATTRIBUTE_TYPE,
+                "Chrome");
+
+        com.optimizely.ab.event.internal.payload.Attribute BotFilteringFeature = new com.optimizely.ab.event.internal.payload.Attribute(
+                ReservedAttributeKey.BOT_FILTERING_ATTRIBUTE.toString(),
+                ReservedAttributeKey.BOT_FILTERING_ATTRIBUTE.toString(),
+                com.optimizely.ab.event.internal.payload.Attribute.CUSTOM_ATTRIBUTE_TYPE,
+                validProjectConfig.getBotFiltering()+"");
+        List<com.optimizely.ab.event.internal.payload.Attribute> expectedUserFeatures;
+
+        if(datafileVersion >= Integer.parseInt(ProjectConfig.Version.V4.toString()))
+            expectedUserFeatures = Arrays.asList(userAgentFeature, feature, BotFilteringFeature);
+        else
+            expectedUserFeatures = Arrays.asList(userAgentFeature, feature);
+
+        LogEvent impressionEvent = builder.createImpressionEvent(validProjectConfig, activatedExperiment, bucketedVariation,
+                userId, attributeMap);
+
+        // verify that request endpoint is correct
+        assertThat(impressionEvent.getEndpointUrl(), is(EventBuilder.EVENT_ENDPOINT));
+
+        EventBatch eventBatch = gson.fromJson(impressionEvent.getBody(), EventBatch.class);
+
+        // verify payload information
+        assertThat(eventBatch.getVisitors().get(0).getVisitorId(), is(userId));
+        assertThat((double) eventBatch.getVisitors().get(0).getSnapshots().get(0).getEvents().get(0).getTimestamp(), closeTo((double)System.currentTimeMillis(), 1000.0));
+        assertFalse(eventBatch.getVisitors().get(0).getSnapshots().get(0).getDecisions().get(0).getIsCampaignHoldback());
+        assertThat(eventBatch.getAnonymizeIp(), is(validProjectConfig.getAnonymizeIP()));
+        assertThat(eventBatch.getProjectId(), is(validProjectConfig.getProjectId()));
+        assertThat(eventBatch.getVisitors().get(0).getSnapshots().get(0).getDecisions().get(0), is(expectedDecision));
+        assertThat(eventBatch.getVisitors().get(0).getSnapshots().get(0).getDecisions().get(0).getCampaignId(),
+                is(activatedExperiment.getLayerId()));
+        assertThat(eventBatch.getAccountId(), is(validProjectConfig.getAccountId()));
+        assertThat(eventBatch.getVisitors().get(0).getAttributes(), is(expectedUserFeatures));
+        assertThat(eventBatch.getClientName(), is(EventBatch.ClientEngine.JAVA_SDK.getClientEngineValue()));
+        assertThat(eventBatch.getClientVersion(), is(BuildVersionInfo.VERSION));
+        assertNull(eventBatch.getVisitors().get(0).getSessionId());
```

Comment:
This is so messy. Can't we do it like in other SDKs. 
Create an exepected Log Event object.  And compare two log event objects. 
Instead of comparing individual items. 
I completely refactored ruby code in this manner. Have a general log event object on the top. In each testcase edit the object accordingly and compare it. 
May be you can refactor all of this later